runtime/devnet/src/extensions/mod.rs

```diff
@@ -0,0 +1,1160 @@
+				match function_id {
+					FuncId::Dispatch => {
+						dispatch::<T, E>(&mut env, version, pallet_index, call_index, params)
+					},
+					FuncId::ReadState => {
+						read_state::<T, E>(&mut env, version, pallet_index, call_index, params)
+					},
+					// TODO
+					FuncId::SendXcm => send_xcm::<T, E>(&mut env),
+				}
+			},
+			Err(e) => Err(e),
+		};
+
+		match result {
+			Ok(_) => Ok(RetVal::Converging(0)),
+			Err(e) => Ok(RetVal::Converging(convert_to_status_code(e, version))),
+		}
+	}
+}
+
+fn dispatch<T, E>(
+	env: &mut Environment<E, BufInBufOutState>,
+	version: u8,
+	pallet_index: u8,
+	call_index: u8,
+	params: Vec<u8>,
+) -> Result<(), DispatchError>
+where
+	T: frame_system::Config<RuntimeOrigin = RuntimeOrigin, RuntimeCall = RuntimeCall>,
+	RuntimeOrigin: From<RawOrigin<T::AccountId>>,
+	E: Ext<T = T>,
+{
+	const LOG_PREFIX: &str = " dispatch |";
+	let call = construct_call(version, pallet_index, call_index, params)?;
+	// Contract is the origin by default.
+	let origin: RuntimeOrigin = RawOrigin::Signed(env.ext().address().clone()).into();
+	dispatch_call::<T, E>(env, call, origin, LOG_PREFIX)
+}
+
+fn dispatch_call<T, E>(
+	env: &mut Environment<E, BufInBufOutState>,
+	call: RuntimeCall,
+	mut origin: RuntimeOrigin,
+	log_prefix: &str,
+) -> Result<(), DispatchError>
+where
+	T: frame_system::Config<RuntimeOrigin = RuntimeOrigin, RuntimeCall = RuntimeCall>,
+	RuntimeOrigin: From<RawOrigin<T::AccountId>>,
+	E: Ext<T = T>,
+{
+	let charged_dispatch_weight = env.charge_weight(call.get_dispatch_info().weight)?;
+	log::debug!(target:LOG_TARGET, "{} Inputted RuntimeCall: {:?}", log_prefix, call);
+	origin.add_filter(AllowedApiCalls::contains);
+	match call.dispatch(origin) {
+		Ok(info) => {
+			log::debug!(target:LOG_TARGET, "{} success, actual weight: {:?}", log_prefix, info.actual_weight);
+			// Refund weight if the actual weight is less than the charged weight.
+			if let Some(actual_weight) = info.actual_weight {
+				env.adjust_weight(charged_dispatch_weight, actual_weight);
+			}
+			Ok(())
+		},
+		Err(err) => {
+			log::debug!(target:LOG_TARGET, "{} failed: error: {:?}", log_prefix, err.error);
+			Err(err.error)
+		},
+	}
+}
+
+fn construct_call(
+	version: u8,
+	pallet_index: u8,
+	call_index: u8,
+	params: Vec<u8>,
+) -> Result<RuntimeCall, DispatchError> {
+	match pallet_index {
+		52 => {
+			let call = versioned_construct_assets_call(version, call_index, params)?;
+			Ok(RuntimeCall::Assets(call))
+		},
+		_ => Err(DispatchError::Other("UnknownFunctionId")),
+	}
+}
+
+fn construct_key(
+	version: u8,
+	pallet_index: u8,
+	call_index: u8,
+	params: Vec<u8>,
+) -> Result<RuntimeStateKeys, DispatchError> {
+	match pallet_index {
+		52 => {
+			let key = versioned_construct_assets_key(version, call_index, params)?;
+			Ok(RuntimeStateKeys::Assets(key))
+		},
+		_ => Err(DispatchError::Other("UnknownFunctionId")),
+	}
+}
+
+fn versioned_construct_assets_call(
```

Comment:
All these boilerplate functions seem unnecessary to me, when we can just have an enum like RuntimeCall which handles all of these decoding in a much more graceful manner. We dont need to be mindful of size as much on the runtime side, and I would favour easier maintenance over this additional complexity with all these mapping functions.

Reply:
I could be missing something however, this has been a long review.

Reply:
This code will not be necessary with #94 so will resolve.